packages/roosterjs-editor-core/lib/corePlugins/utils/InlineEntityHandlers/inlineEntityOnPluginEvent.ts

```diff
@@ -108,25 +107,54 @@ export function inlineEntityOnPluginEvent(event: PluginEvent, editor: IEditor) {
     }
 }
 
-function getDelimiter(entityWrapper: HTMLElement, after: boolean): HTMLElement | undefined {
-    const el = after ? entityWrapper.nextElementSibling : entityWrapper.previousElementSibling;
-    return el && safeInstanceOf(el, 'HTMLElement') && getDelimiterFromElement(el) ? el : undefined;
+function getDelimiters(entityWrapper: HTMLElement): (HTMLElement | undefined)[] {
+    return [entityWrapper.nextElementSibling, entityWrapper.previousElementSibling].map(el =>
+        el && safeInstanceOf(el, 'HTMLElement') && getDelimiterFromElement(el) ? el : undefined
+    );
 }
 
-function removeDelimiters(entityWrapper: HTMLElement): void {
-    let el: HTMLElement | undefined = undefined;
-    if ((el = getDelimiter(entityWrapper, true))) {
-        el.parentElement?.removeChild(el);
-    }
-    if ((el = getDelimiter(entityWrapper, false))) {
-        el.parentElement?.removeChild(el);
-    }
+function addDelimitersIfNeeded(nodes: Element[] | NodeListOf<Element>) {
+    nodes.forEach(node => {
+        if (safeInstanceOf(node, 'HTMLElement') && isReadOnly(getEntityFromElement(node))) {
+            const [delimiterAfter, delimiterBefore] = getDelimiters(node);
+
+            if (!delimiterAfter) {
+                addDelimiterAfter(node);
+            }
+            if (!delimiterBefore) {
+                addDelimiterBefore(node);
+            }
+        }
+    });
 }
 
-function isReadOnly(entity: Entity) {
+function removeNode(el: Node | undefined) {
+    el?.parentElement?.removeChild(el);
+}
+
+function isReadOnly(entity: Entity | null) {
     return (
-        entity.isReadonly &&
+        entity?.isReadonly &&
         !isBlockElement(entity.wrapper) &&
         safeInstanceOf(entity.wrapper, 'HTMLElement')
     );
 }
+
+function removeInvalidDelimiters(nodes: Element[]) {
+    nodes.forEach(node => {
+        const sibling = node.classList.contains(DelimiterClasses.DELIMITER_AFTER)
+            ? node.nextElementSibling
+            : node.previousElementSibling;
+        if (
+            getDelimiterFromElement(node) &&
+            !!(safeInstanceOf(sibling, 'HTMLElement') && getEntityFromElement(sibling))
+        ) {
+            removeNode(node);
+        } else {
+            node?.classList.remove(
+                DelimiterClasses.DELIMITER_BEFORE,
+                DelimiterClasses.DELIMITER_AFTER
+            );
```

Comment:
I wonder if this is okay, this can happen when in a scenario the user was able to type inside of a delimiter and the textContent is no longer ZWS. Not sure if we should also try to remove the ZWS from the text content.

Reply:
It would be good if we can also remove ZWS.